### PR TITLE
Print improvements: don't print unprintable filetypes, improve PDF conversion and mimetype detection

### LINKED ISSUE #918: Disable "Print" when file format is not conducive to printing
## Problem
A bit to file in the "duh" bucket, and thus something I'm not at all worried about until the post-Pilot release...

Ya can't print a video file. Or a song, or a ZIP, an EXE, etc. As such, the obviousness of this qualifies it as a low-priority to me, at the time of this filling. 

image

## Solution
Disable the "Print" text-link next to files that are somehow not printable; be they compressed archives, videos, audio, or any other non-printable dealie-o. On second thought, perhaps best to approach this as only _enabling_ the "Print" button to be live, for supported/known filetypes that can deal with printing. Images less than 100mb (or whatever), .doc, .pdf, .xls, .txt, etc.

### LINKED ISSUE #1725: [securedrop-export] Broaden printable filetypes to include more types
We currently convert Office/Libreoffice files using `unoconv` to print them as PDFs. `unoconv` (already installed in sd-devices) currently supports the following filetypes: [Edit: see comment below regarding moving away from unoconv] 

```
The following list of document formats are currently available:

  bib      - BibTeX [.bib]
  doc      - Microsoft Word 97/2000/XP [.doc]
  doc6     - Microsoft Word 6.0 [.doc]
  doc95    - Microsoft Word 95 [.doc]
  docbook  - DocBook [.xml]
  docx     - Microsoft Office Open XML [.docx]
  docx7    - Microsoft Office Open XML [.docx]
  fodt     - OpenDocument Text (Flat XML) [.fodt]
  html     - HTML Document (OpenOffice.org Writer) [.html]
  latex    - LaTeX 2e [.ltx]
  mediawiki - MediaWiki [.txt]
  odt      - ODF Text Document [.odt]
  ooxml    - Microsoft Office Open XML [.xml]
  ott      - Open Document Text [.ott]
  pdf      - Portable Document Format [.pdf]
  rtf      - Rich Text Format [.rtf]
  stw      - Open Office.org 1.0 Text Document Template [.stw]
  sxw      - Open Office.org 1.0 Text Document [.sxw]
  text     - Text Encoded [.txt]
  txt      - Text [.txt]
  uot      - Unified Office Format text [.uot]
  xhtml    - XHTML Document [.html]
  epub     - Electronic Publication [.epub]

The following list of graphics formats are currently available:

  bmp      - Windows Bitmap [.bmp]
  emf      - Enhanced Metafile [.emf]
  eps      - Encapsulated PostScript [.eps]
  fodg     - OpenDocument Drawing (Flat XML) [.fodg]
  gif      - Graphics Interchange Format [.gif]
  html     - HTML Document (OpenOffice.org Draw) [.html]
  jpg      - Joint Photographic Experts Group [.jpg]
  met      - OS/2 Metafile [.met]
  odd      - OpenDocument Drawing [.odd]
  otg      - OpenDocument Drawing Template [.otg]
  pbm      - Portable Bitmap [.pbm]
  pct      - Mac Pict [.pct]
  pdf      - Portable Document Format [.pdf]
  pgm      - Portable Graymap [.pgm]
  png      - Portable Network Graphic [.png]
  ppm      - Portable Pixelmap [.ppm]
  ras      - Sun Raster Image [.ras]
  std      - OpenOffice.org 1.0 Drawing Template [.std]
  svg      - Scalable Vector Graphics [.svg]
  svm      - StarView Metafile [.svm]
  swf      - Macromedia Flash (SWF) [.swf]
  sxd      - OpenOffice.org 1.0 Drawing [.sxd]
  tiff     - Tagged Image File Format [.tiff]
  wmf      - Windows Metafile [.wmf]
  xhtml    - XHTML [.xhtml]
  xpm      - X PixMap [.xpm]

The following list of presentation formats are currently available:

  bmp      - Windows Bitmap [.bmp]
  emf      - Enhanced Metafile [.emf]
  eps      - Encapsulated PostScript [.eps]
  fodp     - OpenDocument Presentation (Flat XML) [.fodp]
  gif      - Graphics Interchange Format [.gif]
  html     - HTML Document (OpenOffice.org Impress) [.html]
  jpg      - Joint Photographic Experts Group [.jpg]
  met      - OS/2 Metafile [.met]
  odg      - ODF Drawing (Impress) [.odg]
  odp      - ODF Presentation [.odp]
  otp      - ODF Presentation Template [.otp]
  pbm      - Portable Bitmap [.pbm]
  pct      - Mac Pict [.pct]
  pdf      - Portable Document Format [.pdf]
  pgm      - Portable Graymap [.pgm]
  png      - Portable Network Graphic [.png]
  potm     - Microsoft PowerPoint 2007/2010 XML Template [.potm]
  pot      - Microsoft PowerPoint 97/2000/XP Template [.pot]
  ppm      - Portable Pixelmap [.ppm]
  pptx     - Microsoft PowerPoint 2007/2010 XML [.pptx]
  pps      - Microsoft PowerPoint 97/2000/XP (Autoplay) [.pps]
  ppt      - Microsoft PowerPoint 97/2000/XP [.ppt]
  pwp      - PlaceWare [.pwp]
  ras      - Sun Raster Image [.ras]
  sda      - StarDraw 5.0 (OpenOffice.org Impress) [.sda]
  sxd      - OpenOffice.org 1.0 Drawing (OpenOffice.org Impress) [.sxd]
  sti      - OpenOffice.org 1.0 Presentation Template [.sti]
  svg      - Scalable Vector Graphics [.svg]
  svm      - StarView Metafile [.svm]
  swf      - Macromedia Flash (SWF) [.swf]
  sxi      - OpenOffice.org 1.0 Presentation [.sxi]
  tiff     - Tagged Image File Format [.tiff]
  uop      - Unified Office Format presentation [.uop]
  wmf      - Windows Metafile [.wmf]
  xhtml    - XHTML [.xml]
  xpm      - X PixMap [.xpm]

The following list of spreadsheet formats are currently available:

  csv      - Text CSV [.csv]
  dbf      - dBASE [.dbf]
  dif      - Data Interchange Format [.dif]
  fods     - OpenDocument Spreadsheet (Flat XML) [.fods]
  html     - HTML Document (OpenOffice.org Calc) [.html]
  ods      - ODF Spreadsheet [.ods]
  ooxml    - Microsoft Excel 2003 XML [.xml]
  ots      - ODF Spreadsheet Template [.ots]
  pdf      - Portable Document Format [.pdf]
  slk      - SYLK [.slk]
  stc      - OpenOffice.org 1.0 Spreadsheet Template [.stc]
  sxc      - OpenOffice.org 1.0 Spreadsheet [.sxc]
  uos      - Unified Office Format spreadsheet [.uos]
  xhtml    - XHTML [.xhtml]
  xls      - Microsoft Excel 97/2000/XP [.xls]
  xls5     - Microsoft Excel 5.0 [.xls]
  xls95    - Microsoft Excel 95 [.xls]
  xlt      - Microsoft Excel 97/2000/XP Template [.xlt]
  xlt5     - Microsoft Excel 5.0 Template [.xlt]
  xlt95    - Microsoft Excel 95 Template [.xlt]
  xlsx     - Microsoft Excel 2007/2010 XML [.xlsx]
```
It would be easy to add support for more of these filetypes (see eg https://github.com/freedomofpress/securedrop-export/pull/109), would be helpful for users (see eg https://github.com/freedomofpress/securedrop-client/issues/2008 or https://github.com/freedomofpress/securedrop-client/issues/2007; though not about printing, the same concept applies), and would not require adding any additional dependencies. Let's discuss which formats are useful to include, as a stopgap while we figure out other formats such as webp support, or while we figure larger changes such as https://github.com/freedomofpress/securedrop-workstation/issues/842.

Note: `unoconv` is also installed in sd-viewer, if we are looking for a quick way to add additional viewable filetypes without additional upstream dependencies.

### PR DESCRIPTION
**Pull request by rocodes**
## Status

Ready for review 

Documentation changes tk as well

## Description

Printer-focused UX improvements/support: (#918)
- Detect file mimetypes via `mimetype` instead of by file extension before printing
- Add new status codes to report mimetype detection/conversion/unprintable types back to the user
- Update print dialog to display a message if the user tries to print an unprintable type

Broadening printable filetype support: (#1725) 
- Use libreoffice headless mode instead of unoconv for file conversion
- Attempt to convert for printing any file type that is supported by libreoffice, by parsing its .desktop entries on sd-devices. This broadens our printable filetypes from ~10 to ... [a lot more than 10](https://github.com/LibreOffice/core/blob/master/solenv/inc/mime.types) :) ... although some may not print well.  (This isn't quite an accurate list; I will add a docs PR). 

Dependency management:
- Specify libreoffice as a dependency of the securedrop-export and securedrop-viewer packages in the debian/control file. This is a change for two reasons:
    - Removes the `no-install-recommends` flag, meaning Java (openjdk-17-jdk at present time) will be installed, which it currently is not. This is intentional; it will allow us to support viewing and converting more filetypes. When updated client packages have been released, we can remove the Salt logic that installs libreoffice: see https://github.com/freedomofpress/securedrop-workstation/pull/1162   
- Adjust paxctld.conf to allow openjdk-17 (needed for a dependency of libreoffice, ca-certificates-java). 
- Deprecate unoconv dependency

Fixes #918
Fixes #1725  (although we will still want to look at an overall consistent way of handling file types in the client, and there have been a few suggestions)
Refs (probably fixes, but needs testing) https://github.com/freedomofpress/securedrop-client/issues/1731

## Test Plan

- [ ] Visual review
- [ ] CI passing
- [ ] Manual testing (below)

#### Manual testing 
- You will need a supported USB printer.
- Start from SecureDrop Workstation 1.0.0 rpm or tag and provision SDW. Build packages from the tip of this branch, and install them in the large template. You will need to install at minimum securedrop-client, securedrop-export, securedrop-workstation-viewer, and securedrop-workstation-config.
- [x] Packages install successfully
- Shut down the template, then start SDW. You will need to submit various file types to your instance for testing: please submit: a multi-page PDF; a plain .txt file; one or more video/audio files; one or more image files; one file with the wrong extension; and one or more files supported by libreoffice.
**Viewing**
- [ ] Files that are currently supported by securedrop-workstation viewer can be viewed in viewer Dispvm (no change, no added filetype support, just regression testing)   
**Printing**
- [x] Printing a multi-page PDF is successful
- [x] Printing a .txt file is successful
- [x] Printing a non-PDF file is successful; you can choose from any of the filetypes supported by libreoffice ( :open_mouth:  ), although definitions of "success" may be stretched for some filetypes.
- [x] Attempting to print a video or audio file or any other "unprintable" mimetype (see below) results in the printer modal error that says "This file type cannot be printed" (#918). You don't have to try every single type.
<details><summary>unprintable types</summary>
    "audio/mp4",
    "audio/mpeg",
    "audio/x-vorbis",
    "audio/x-wav",
    "video/quicktime",
    "video/x-theora",
    "video/mp4",
    "video/webm",
    "video/x-msvideo",
    "video/x-ms-wmv",
    "application/vnd.djvu",
    "application/vnd.rar",
    "application/zip",
    "application/x-7z-compressed"
     </details>

## Checklist

TK

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [ ] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [x] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration is [self-contained] and applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [x] No database schema changes are needed

[self-contained]: https://github.com/freedomofpress/securedrop-client#generating-and-running-database-migrations

### CODE REVIEW

File: export/securedrop_export/print/service.py
Comment (legoktm) [inline export/securedrop_export/print/service.py:20]: Can we prefix match all of `audio/` and `video/`?

File: export/securedrop_export/print/service.py
Comment (legoktm) [inline export/securedrop_export/print/service.py:246]: This is cool :)

File: export/securedrop_export/print/service.py
Comment (legoktm) [inline export/securedrop_export/print/service.py:293]: Any difference in using `mimetype` vs `file --mime` or https://pypi.org/project/python-magic/ ?

**Review by legoktm — COMMENTED** [review]
I just did a quick skim!

File: export/securedrop_export/print/service.py
Comment (rocodes) [reply export/securedrop_export/print/service.py]: I ended up going with `mimetype -M` (the magic bytes flag). 
Mimetype looks first in XDG_DATA_DIRS (and other places), and `file` looks in /etc/mime I think.

**Comment by rocodes** [comment]
I was hoping to have this all finished + green before tomorrow and to do a little more testing, but I'm a little short on time and still need to fix up one or two tests, but I'm taking it out of draft mode to invite more feedback and testing. So far I've tested the export component with locally-built packages, and printing and mime detection are working as expected.

I will be back to address review feedback (such as https://github.com/freedomofpress/securedrop-client/pull/2166#discussion_r1718652578 and anything else that arises) and do test fixup on Wed.

File: export/tests/print/test_service.py
Comment (legoktm) [inline export/tests/print/test_service.py:136]: Since we're testing against fixed sample files, can we assert the real number? Or even just the full list of mime types?

File: export/securedrop_export/print/service.py
Comment (legoktm) [inline export/securedrop_export/print/service.py:268]: nit: You could avoid some extra indentation with `for line in desktop_file.read_text().splitlines()`.

**Review by legoktm — COMMENTED** [review]
Excited for this, I've done a code review pass and now am going to actually test it.

**Comment by legoktm** [comment]
Test plan checks out, one note: I tried "one file with the wrong extension" by having a webp image with the .txt extension, it opened in the wrong application (text editor instead of eog), I don't think that's a regression here right?

**Review by legoktm — CHANGES_REQUESTED** [review]
Test plan checks out, one note: I tried "one file with the wrong extension" by having a webp image with the .txt extension, it opened in the wrong application (text editor instead of eog), I don't think that's a regression here right?

Marking as requesting changes based on previous CR

File: export/tests/print/test_service.py
Comment (legoktm) [inline export/tests/print/test_service.py:354]: This is failing in CI because it should be mocking `pathlib.Path.exists`, but then it fails because there are other Path.exists() calls that get invoked and some of them should return False (e.g. checking if the PDF file already exists).

**Comment by legoktm** [comment]
I pushed https://github.com/freedomofpress/securedrop-client/commit/292b23010f14c47a4a74cfd58ef304cd1fe983b3 in a separate branch to show how we can install LibreOffice in CI and use it for integration tests; there could be others that run various files through headless libreoffice to see how the conversion goes and error handling instead of trying to mock it all out.

Let me know if you'd like me to push it to this branch or submit it as a separate PR after this lands! (Or feel free to pull it in yourself :))

**Comment by rocodes** [comment]
Hey, thank you for all the feedback and testing :)  Realquick:
- These changes don't affect sd-viewer (and don't broaden viewing support - only printing support), so you're right that the .txt file being opened in the wrong application is not related (but we should fix it).
- I have local changes to address a couple failures including the pathlib thing you noticed, but I ran out of time and didn't want to push anything in haste - sorry if you spent too much time looking into that.
- Thanks for all your cleanup changes - much better - shall include them too.  And I'll take a look at your libreoffice branch. I was on the fence about including that in CI or whether that blurs the boundaries between unit-testing and integration testing, but I'm into it.

**Comment by legoktm** [comment]
> * so you're right that the .txt file being opened in the wrong application is not related (but we should fix it).

Ack, filed separately as #2171.

> * I have local changes to address a couple failures including the pathlib thing you noticed, but I ran out of time and didn't want to push anything in haste - sorry if you spent too much time looking into that.

OK! I was going to spend some time today figuring that out but I can hold off for now then.

> * And I'll take a look at your libreoffice branch. I was on the fence about including that in CI or whether that blurs the boundaries between unit-testing and integration testing, but I'm into it.

It definitely does blur the boundaries, but I think mixing is fine, if we end up having more we can split them out into separate files.

Also, integration tests are important:
eb6fe1b73f87abe806b353aacbde464c

**Comment by legoktm** [comment]
test are failing because it doesn't find any libreoffice .desktop mime types:
```
        if len(supported_types) == 0:
>           raise ExportException(sdstatus=Status.ERROR_MIMETYPE_DISCOVERY)
```

I suspect you (like me) have LibreOffice installed locally so it passes locally but the CI container doesn't have it, so it either needs to be mocked or we just install LibreOffice in CI and it works (which it will once my integration patch is in)

**Comment by rocodes** [comment]
:) I did catch that, and I _think_ I addressed all your changes too in [05fe7dd](https://github.com/freedomofpress/securedrop-client/pull/2166/commits/05fe7dd12cc7e3bfb32b870e8800aaed507774b3)..https://github.com/freedomofpress/securedrop-client/pull/2166/commits/706e94a3a095e4436185c719e6141082f472627c

**Comment by rocodes** [comment]
(moving out of 'ready for review' until I pull in https://github.com/freedomofpress/securedrop-client/commit/292b23010f14c47a4a74cfd58ef304cd1fe983b3)

File: export/securedrop_export/print/service.py
Comment (legoktm) [inline export/securedrop_export/print/service.py:306]: A cool trick is that `startswith` takes a tuple of strings, which allows you to do:
```
>>> 'audio/foo'.startswith(('audio/', 'video/'))
True
>>> 'text/foo'.startswith(('audio/', 'video/'))
False
```

**Review by legoktm — APPROVED** [review]
LGTM, in real testing, it all works too!

**Pull request merged**
